File: leetcode/124-binary-tree-maximum-path-sum.cpp
/*
Given a binary tree, find the maximum path sum.

The path may start and end at any node in the tree.

For example:
Given the below binary tree,

       1
      / \
     2   3
Return 6.
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode *root) {
        auto p = _maxPathSum(root);
        return max(p.first, p.second);
    }

private:
    // expandable, non-ex
    pair<int, int> _maxPathSum(TreeNode* root) {
        if (root == nullptr) {
            return {0, 0};
        }
        int exp = root->val;
        int nonexp = root->val;
        pair<int, int> left, right;
        if (root->left != nullptr) {
            left = _maxPathSum(root->left);
            exp = max(exp, root->val + left.first);
            nonexp = max(nonexp, left.second);
            nonexp = max(nonexp, left.first);
        }
        if (root->right != nullptr) {
            right = _maxPathSum(root->right);
            exp = max(exp, root->val + right.first);
            nonexp = max(nonexp, right.second);
            nonexp = max(nonexp, right.first);
        }
        if (root->right != nullptr && root->left != nullptr) {
            nonexp = max(nonexp, right.first + root->val + left.first);
        }
        return {exp, nonexp};
    }
};

File: leetcode/124-binary-tree-maximum-path-sum.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer}
def max_path_sum(root)
  return 0 if root.nil?
  _max_sum(root).select { |x| not x.nil? }.max  
end

# [can_expand, no_expand]
def _max_sum(root)
  return [nil, nil] if root.nil?
  left_e, left_ne = _max_sum root.left
  right_e, right_ne = _max_sum root.right
  e = root.val + [left_e, right_e, 0].select { |x| not x.nil? }.max
  all_ne = (left_e.nil? ? 0 : left_e) + (right_e.nil? ? 0 : right_e) + root.val
  ne = [left_ne, right_ne, left_e, right_e, all_ne].select { |x| not x.nil? }.max
  [e, ne]
end
